pkg/ebpf/dns_stats.go

```diff
@@ -34,36 +38,89 @@ type dnsPacketInfo struct {
 	type_         DNSPacketType
 }
 
+type stateKey struct {
+	key dnsKey
+	id  uint16
+}
+
 type dnsStatKeeper struct {
-	mux   sync.Mutex
-	stats map[dnsKey]dnsStats
+	mux              sync.Mutex
+	stats            map[dnsKey]dnsStats
+	state            map[stateKey]time.Time
+	expirationPeriod time.Duration
+	exit             chan struct{}
+	maxSize          int // maximum size of the state map
+	deleteCount      int
+	deleteThreshold  int
 }
 
-func newDNSStatkeeper() *dnsStatKeeper {
-	return &dnsStatKeeper{
-		stats: make(map[dnsKey]dnsStats),
+func newDNSStatkeeper(timeout time.Duration) *dnsStatKeeper {
+	statsKeeper := &dnsStatKeeper{
+		stats:            make(map[dnsKey]dnsStats),
+		state:            make(map[stateKey]time.Time),
+		expirationPeriod: timeout,
+		exit:             make(chan struct{}),
+		maxSize:          10000,
+		deleteThreshold:  5000,
 	}
+
+	ticker := time.NewTicker(statsKeeper.expirationPeriod)
+	go func() {
+		for {
+			select {
+			case now := <-ticker.C:
+				statsKeeper.removeExpiredStates(now.Add(-statsKeeper.expirationPeriod))
+			case <-statsKeeper.exit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	return statsKeeper
 }
 
-func (d *dnsStatKeeper) ProcessPacketInfo(info dnsPacketInfo) {
+func (d *dnsStatKeeper) ProcessPacketInfo(info dnsPacketInfo, ts time.Time) {
 	d.mux.Lock()
 	defer d.mux.Unlock()
-	stats := d.stats[info.key]
+	sk := stateKey{key: info.key, id: info.transactionID}
 
-	// For local DNS traffic, sometimes the same reply packet gets processed by the
-	// snooper multiple times. This check avoids double counting in that scenario
-	// assuming the duplicates are not interleaved.
-	if stats.lastTransactionID == info.transactionID {
+	if info.type_ == Query {
+		if len(d.state) == d.maxSize {
+			return
+		}
+
+		if _, ok := d.state[sk]; !ok {
+			d.state[sk] = ts
+		}
 		return
 	}
 
-	if info.type_ == SuccessfulResponse {
-		stats.successfulResponses++
-	} else {
-		stats.failedResponses++
+	// If a response does not have a corresponding query entry, we discard it
+	start, ok := d.state[sk]
+
+	if !ok {
+		return
 	}
 
-	stats.lastTransactionID = info.transactionID
+	delete(d.state, sk)
+	d.deleteCount++
+
+	latency := ts.Sub(start).Nanoseconds()
```

Comment:
we're run into memory and CPU problems in the past keeping large number of time.Time and using Sub/Add instead of integer arithmetic operators. Can we  have the values of `d.state` be microseconds, convert `ts` to micros, and do all the time logic with arithmetic ?